src/sage/combinat/integer_vector.py

```diff
@@ -898,6 +901,83 @@ def __contains__(self, x):
             return False
         return sum(x) == self.n
 
+    def rank(self, x):
+        """
+        Return the rank of a given element.
+
+        INPUT:
+
+        - ``x`` -- a list with ``sum(x) == n``
+
+        EXAMPLES::
+
+            sage: IntegerVectors(n=5).rank([5,0])
+            1
+            sage: IntegerVectors(n=5).rank([3,2])
+            3
+        """
+        if sum(x) != self.n:
+            raise ValueError("argument is not a member of IntegerVectors({},{})".format(self.n, None))
+
+        n, k = self.n, len(x)
+
+        r = binomial(k+n-1, n+1)
+
+        for i in range(k - 1):
+            k -= 1
+            n -= x[i]
+            r += binomial(k + n - 1, n - 1)
+
+        return r
+
+    def unrank(self, x):
+        """
+        Return the element at given rank x.
+
+        INPUT:
+
+        - ``x`` -- an integer.
+
+        EXAMPLES::
+
+            sage: IntegerVectors(n=5).unrank(2)
+            [4, 1]
+            sage: IntegerVectors(n=10).unrank(10)
+            [1, 9]
+        """
+        ptr = 0
+        rtn = [self.n]
+        while self.rank(rtn) < x:
+            rtn.append(0)
+        rtn.pop()
+
+        while True:
```

Comment:
maybe this while loop can be factored out into a helper function in `IntegerVectors`?

In any case, we should store `current_rank = self.rank(rtn)`, so we don't compute it twice.

Reply:
I think no. If I see it correctly, it would only be 3 lines there.